### ISSUE TITLE
Implement Rich Types

### ISSUE THREAD

**Issue by mfcastellani** [mfcastellani/hcl-checker#11]
Need to implement Rich Types, available at Terraform 0.12

Terraform providers and resources always expect specific data types to be provided. Terraform will continue to attempt to automatically convert a provided value to the expected type if it doesn't match. An error will be produced if this isn't possible.

Terraform 0.12 introduces a powerful type system that module authors can use to specify expected types for input variables. This forces values to be validated and converted properly if possible. This also improves the error messages shown to users of a module.

Terraform has always supported a very simple type requirement for variables. The example below uses Terraform 0.12 syntax, but showcases functionality that has been possible in Terraform for many years:

#### Configuration for Terraform 0.12

```
variable "environment_name" {
  type = string
}
```

Experienced Terraform users may notice that a difference is that string is not quoted. Types are now first-class values in Terraform and can be used directly.

Terraform 0.12 can now also represent the types of complex object structures like those shown in the previous section. This was not possible in earlier versions of Terraform. The example below shows a type specification for the example above:

#### Configuration for Terraform 0.12

```
variable "networks" {
  type = map(object({
    network_number    = number
    availability_zone = string
    tags              = map(string)
  }))
}
```

This results in great validation for users of a Terraform configuration or module with clear error messages.
